3836/3836.md

```diff
@@ -0,0 +1,666 @@
+value. The value is used as the starting point in a carefully-constructed but
+somewhat-brittle blockchain generation procedure relying heavily on `fix` and
+lazy evaluation, and I am taken to believe that these errors are overwritten or
+otherwise left unevaluated with standard usage of this function. (For example, it
+seems that results of `mkPartialBlockHeader`, the enclosing function, are passed
+immediately to `fixupBlock`, which overwrites the stubbed-out fields.)
+
+I think it would be possible to refactor this chain generation to avoid having
+to utter `error`, but it would not be a quick fix, and the testing-based nature
+and use of this package leads me to believe that such a fix may not provide
+substantial value.
+
+Regardless, I've changed the error message to describe the function enclosing
+it, which is `mkPartialBlockHeader`, not `mkPartialBlock`.
+
+**Fix: message**
+#fixed
+**Fix: refactor**
+#unfixed
+
+#input-desired
+
+
+### 7: `ouroboros-network-mock/src/Ouroboros/Network/Mock/Chain.hs:235`
+```hs
+go _ = error "successorBlock: point not on chain"
+```
+
+The enclosing function, `successorBlock`, tries to find the successor block of a
+given point on a given chain. At the moment, this error is triggered iff the
+point is not on the chain.
+
+All callers already check point membership on the chain, via `pointOnChain`,
+before calling this function. Additionally, the QuickCheck test of this
+function's behavior is structured as an implication, where use of
+`successorBlock` is guarded by a positive result from `pointOnChain`. I
+therefore claim that an idiom has emerged: using `successorBlock` to find the
+successor of a nonexistent point is considered undefined behavior. I propose to
+codify this idiom in the function's contract, then redefine the undefined
+behavior to be `Nothing`, a non-`error` result that happens to inhabit the
+function's return type.
+
+**Fix: refactor**
+#fixed
+
+#input-desired
+
+
+### 8/9: `ouroboros-network-testing/src/Ouroboros/Network/Testing/Data/Script.hs:221/223`
+```hs
+= error "interpretPickScript: given empty map to pick from"
+-- ...
+= error "interpretPickScript: given invalid pickNum"
+```
+
+`interpretPickScript` is used to create `PickPolicy`s to initialize a
+`PeerSelectionPolicy`. Its two errors both ensure the same precondition that
+`Ouroboros.Network.PeerSelection.Governor.Types.pickPeers` ensures when
+executing a `PickPolicy` - that there are peers available to pick from, and that
+we are picking more than zero of them.
+
+It's my understanding that `PickPolicy`s aren't called directly, but are always
+invoked via `pickPeers`. If this is true, then we can remove these error checks
+from `interpretPickScript` entirely, trusting that the checks in `pickPeers`
+will catch the same failures. If this isn't true, then we could either leave
+this code as-is or change it to the `assert` style favored by `pickPeers`.
+
+**Fix: uncertain**
+#unfixed
+
+#input-desired
```

Comment:
I agree, if they are redundant we can remove them.

Reply:
You commented after this, inline in the file in question, that we might ought to leave these checks in place, so I'll assume that later opinion governs.